lib/spack/spack/binary_distribution.py

```diff
@@ -133,6 +131,349 @@ def __init__(self, errors):
+        self.spec_dict: Dict[Any, Any] = {}
+
+        self._checked_signed = False
+        self._checked_unsigned = False
+        self._checked_tarball = False
+
+    def get_layout_version(self):
+        return self.layout_version
+
+    def initialize_from_spec_and_mirror(self, spec: spack.spec.Spec, mirror_url: str):
+        if not spec.concrete:
+            raise BuildcacheEntryError("Concrete spec required for URLBuildcacheEntry")
+
+        self.mirror_url = mirror_url
+        self.remote_spec_url_signed = self.compute_remote_spec_url(spec, mirror_url, signed=True)
+        self.remote_spec_url_unsigned = self.compute_remote_spec_url(
+            spec, mirror_url, signed=False
+        )
+
+    def initialize_from_spec_url(self, spec_url: str):
+        """Initialized the buildcache entry from a url to a spec metadata file.  If the
+        metadata url is not for a v3 spec, or otherwise does match the expected format
+        or does not exist, an exception is raised."""
+        rematch = URLBuildcacheEntry.SPEC_URL_REGEX.search(spec_url)
+
+        if not rematch:
+            raise BuildcacheEntryError(f"Provided url does not match expected format: {spec_url}")
+
+        remote_layout_version = int(rematch.group(2))
+        if not remote_layout_version == 3:
+            raise BuildcacheEntryError(f"{spec_url} is not a v3 spec")
+
+        self.mirror_url = rematch.group(1)
+
+        if not web_util.url_exists(spec_url):
+            raise BuildcacheEntryError(f"No spec could be found at the given url: {spec_url}")
+
+        self.remote_spec_url = spec_url
+
+        if self.remote_spec_url.endswith(".sig"):
+            self.has_signed = True
+            self.has_unsigned = False
+            self.remote_spec_url_signed = spec_url
+            self.remote_spec_url_unsigned = spec_url[:-4]
+        else:
+            self.has_signed = False
+            self.has_unsigned = True
+            self.remote_spec_url_unsigned = spec_url
+            self.remote_spec_url_signed = f"{spec_url}.sig"
+
+    def get_base_url(self, spec_url: str) -> str:
+        rematch = URLBuildcacheEntry.SPEC_URL_REGEX.search(spec_url)
+        if not rematch:
+            raise BuildcacheEntryError(f"Unable to parse spec url: {spec_url}")
+        return rematch.group(1)
+
+    def compute_remote_archive_url(self, mirror_url: str, algorithm: str, checksum: str) -> str:
+        rel_tarball_components = self.get_relative_tarball_components(algorithm, checksum)
+        return url_util.join(mirror_url, *rel_tarball_components)
+
+    def compute_remote_spec_url(
+        self, spec: spack.spec.Spec, mirror_url: str, signed: bool = True
+    ) -> str:
+        url_prefix = url_util.join(mirror_url, *self.get_relative_spec_components(spec, ".spec"))
+        ext = ".json.sig" if signed else ".json"
+        return f"{url_prefix}{ext}"
+
+    def _check_metadata_exists(self):
+        if not self._checked_signed:
```

Comment:
This is going to re-check for either signed or unsigned every time it is called. Also, could this early exit if `self.spec_dict` is set?

I don't think we support both signed and unsigned metadata in the cache, so we can probably exit once one is found.

Reply:
We always [checked](https://github.com/spack/spack/blob/develop/lib/spack/spack/binary_distribution.py#L1106-L1108) for both before, and it just does a HEAD request, so I think it's not very expensive.  Still, I think it makes sense to only check for unsigned if we didn't have signed, so I've done that.

Regarding exit early: this starts with `_` and isn't used except from `fetch_metadata()`, where the check for `self.spec_dict` is done before calling this method. Do you still think it should be checked here?  Or checked here instead?

Reply:
I think checking here where it is relevant makes sense.

On we always checked both, that is one of the parts of signing/build cache layout I am targeting to fix in `v4`. I am not going to block in this though since it is consistent with the old behavior.

Reply:
Ok, possibly something for the future then.